components/teams-service/storage/postgres/postgres.go

```diff
@@ -179,29 +193,65 @@ func (p *postgres) EditTeam(ctx context.Context, team storage.Team) (storage.Tea
 
 // EditTeamByName edits the team's description and projects by name
 func (p *postgres) EditTeamByName(ctx context.Context,
-	teamName string, teamDescription string, teamProjects []string) (storage.Team, error) {
+	teamName string, teamDescription string, newProjects []string) (storage.Team, error) {
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	if newProjects == nil {
+		newProjects = []string{}
+	}
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return storage.Team{}, p.processError(err)
+	}
 
 	projectsFilter, err := ProjectsListFromContext(ctx)
 	if err != nil {
 		return storage.Team{}, p.processError(err)
 	}
 
-	var t storage.Team
-	if teamProjects == nil {
-		teamProjects = []string{}
+	var oldProjects []string
+	err = tx.QueryRowContext(ctx,
+		`SELECT t.projects FROM teams t
+		WHERE t.name = $1 AND projects_match(t.projects, $2::TEXT[])
+		FOR UPDATE;`,
+		teamName, pq.Array(projectsFilter)).
+		Scan(pq.Array(&oldProjects))
+	if err != nil {
+		return storage.Team{}, p.processError(err)
 	}
 
-	err = p.db.QueryRowContext(ctx,
+	projectDiff := projectassignment.CalculateProjectDiff(oldProjects, newProjects)
+	if len(projectDiff) != 0 {
+		// will only return an error if authz is in v2.1 mode
+		_, err := p.authzClient.ValidateProjectAssignment(ctx, &authz_v2.ValidateProjectAssignmentReq{
+			Subjects:   auth_context.FromContext(auth_context.FromIncomingMetadata(ctx)).Subjects,
+			ProjectIds: projectDiff,
```

Comment:
Since we are passing the `projectDiff`, which includes deleted projects, we are potentially returning a `not found` in the event that that project is already deleted. Do we care?

Reply:
That's one of the reasons we have our new cereal ticket! The case where a team contains a deleted project shouldn't happen if we are doing things properly. Even with cereal though, there is still a race condition between the project getting tombstoned and it getting deleted from the teams domain. In that case, I'd be fine with just aborting the request here. It should be a very small race condition. Would need to think harder about what the right thing to do for that edge case is.

Reply:
we originally discussed not checking perms for deletion (but I'm personally on board to do it)

Reply:
We need to check permissions for removing projects but maybe shouldn't fail if they don't exist.